### ISSUE TITLE
Race condition in 09-network's tests

### ISSUE THREAD

**Issue by stainlesspot** [fmi/java-course#40]
Several runs on the given tests in [09-network](https://github.com/fmi/java-course/blob/master/09-network/lab/test/bg/sofia/uni/fmi/java/network/server/nio/CommandExecutionServerTest.java) reveal that (randomly) some of them sometimes fail and sometimes don't.
I suspect the problem comes from the fact that the server is on a different thread, thus a test may try to connect to the server before it has started.

The solution would be making each test wait for the server untill it has started. But I fail to find a painless and elegant way to do this in the current setup. How could this problem be fixed?

**Comment by stainlesspot** [comment]
One solution I see is to add to add a new field to `CommandExecutionServer`:
```java
private boolean started = false;
```
which is updated to true when the server actually starts.

And a new method to `CommandExecutionServer`:
```java
public void started() {
    while (!started) {
        Thread.sleep(200);
   }
}
```
And then in the tests:
```java
@Before
public void waitForServer() {
    commandExServer.started();
}
```

But I would prefferably avoid `Thread.sleep`.

**Comment by 100yo** [comment]
Thanks for noting that.

I was not able to reproduce locally this issue - start of @Test methods execution before the server launched in setUpBeforeClass() has started, although it is not impossible to happen. I have pushed a patch to add a Thread.sleep loop in the setUpBeforeClass() method, to make sure the server has started (by successfully connecting to it).

This solutions seems the cleanest as it does not change the implementation but just the test. Thread.sleep is ugly anyway but I cannot think of a solution without sleeping. This is the only thing that clients could do in case they try to connect to a server before it has started. They do not have a reference to the server object (and could not have - they run in a different JVM, potentially running on a different physical machine), so the option you propose with the waitForServer() would not be possible.

One smart solution for similar synchronization challenge - wait in a thread until certain work has been done by other thread(s) (but the thread(s) did not complete so that you could join() them) is using a CountDownLatch from java.util.concurrent. The CountDownLatch instance must be shared among the threads that need synchronization. Upon creation, its counter is initialized with the number of threads you want to wait, and you wait on the latch. Each thread decrements the latch's counter when it finishes its job. When the counter reached 0, the thread that has waited on it automatically continues its execution. Unfortunately, this beautiful solution of course works only within one JVM, but it is a great synchronization pattern to know.

Check it out: https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/CountDownLatch.html

By the way, it was worth having another look at this particular test because it was not JUnit 4 - compatible, so it needed a fix anyway. You get a bonus point!:)

**Comment by ialidzhikov** [comment]
Fixed by dc1e812
